### PR TITLE
Simplify subsampling

### PR DESCRIPTION
**Pull request by turian**
PLEASE REVIEW CAREFULLY.

* split_key is added back, and used for train/val/test splitting.
* `SubsampleSplit` is rewritten much more simply, as is the sampler.

### CODE REVIEW

File: heareval/tasks/pipeline.py [hunk 1]
```diff
@@ -10,21 +10,13 @@
 from typing import Dict, List, Optional, Set, Union
 from urllib.parse import urlparse
 
+import heareval.tasks.util.audio as audio_util
 import luigi
 import pandas as pd
-from pandas import DataFrame, Series
+from heareval.tasks.util.luigi import WorkTask, download_file, new_basedir
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:16]: What's up with these import moves - the order was correct before https://www.python.org/dev/peps/pep-0008/#imports

File: heareval/tasks/pipeline.py [hunk 2]
```diff
@@ -447,50 +418,40 @@ def requires(self):
...
-            )
-            print(
-                "Datapoints in split after resampling: "
-                f"{len(sampled_subsample_metadata)}"
-            )
-            assert perform_metadata_subsampling(
-                subsample_metadata.sample(frac=1), max_files
-            ).equals(sampled_subsample_metadata), "The subsampling is not stable"
+            subsampled_relpaths = set(relpaths[:max_files])
+            print(f"Files in split after resampling: f{len(subsampled_relpaths)}")
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:440]: Stay f in the string right before the {} -- would be nice to include the split name as well (already mentioned in previous PR)

File: heareval/tasks/pipeline.py [hunk 3]
```diff
@@ -191,50 +195,42 @@ def get_all_metadata(self) -> pd.DataFrame:
...
         It should be unique, it should be obvious what the original
         filename was, and perhaps it should contain the label for
         audio scene tasks.
-        You can override this and simplify if the slugified filename
-        for this dataset is too long.
-
-        The basic version here takes the filename and slugifies it.
         """
         slug_text = str(Path(relative_path).stem)
         slug_text = slug_text.replace("-", "_negative_")
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:214]: How did this end up in the the main pipeline for this task? Maybe I missed this in a previous PR. This replacement was specific for dcase. We don't really want to blanket replace every "-" with "_negative_" -- would rather check for "-" and raise an error here.

File: heareval/tasks/speech_commands.py [hunk 4]
```diff
@@ -7,16 +7,14 @@
 from pathlib import Path
 from typing import List
 
+import heareval.tasks.pipeline as pipeline
+import heareval.tasks.util.luigi as luigi_util
```
Comment (jorshi) [inline heareval/tasks/speech_commands.py:11]: Same with import ordering here, it was correct before

File: heareval/tasks/pipeline.py [hunk 5]
```diff
@@ -447,50 +418,40 @@ def requires(self):
...
-                subsample_metadata, max_files
-            )
-            print(
-                "Datapoints in split after resampling: "
-                f"{len(sampled_subsample_metadata)}"
-            )
-            assert perform_metadata_subsampling(
-                subsample_metadata.sample(frac=1), max_files
-            ).equals(sampled_subsample_metadata), "The subsampling is not stable"
+            subsampled_relpaths = set(relpaths[:max_files])
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:439]: Why set? relpaths are already unique.

File: heareval/tasks/pipeline.py [hunk 6]
```diff
@@ -447,50 +418,40 @@ def requires(self):
...
-            .drop_duplicates(subset="subsample_key", ignore_index=True)
-        )
-        return subsample_metadata
-
     def run(self):
-        subsample_metadata = self.get_subsample_metadata()
-        num_files = len(subsample_metadata)
+        split_metadata = self.metadata[self.metadata["split"] == self.split]
+        relpaths = split_metadata["relpath"].unique()
+        rng = random.Random("SubsampleSplit")
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:424]: Haha, nice seed

File: heareval/tasks/pipeline.py [hunk 7]
```diff
@@ -364,36 +360,11 @@ def run(self):
...
-            ).issubset(set(metadata.columns))
+        if self.task_config["embedding_type"] == "scene":
             # Multiclass predictions should only have a single label per file
             if self.task_config["prediction_type"] == "multiclass":
                 label_count = metadata.groupby("slug")["label"].aggregate(len)
                 assert (label_count == 1).all()
-        else:
-            raise ValueError(
-                "%s embedding_type unknown" % self.task_config["embedding_type"]
-            )
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:396]: Would be good to still include an assert on the embedding type somewhere?

**Review by jorshi — APPROVED** [review]
Couple lil comments, but overall this simplification looks great

File: heareval/tasks/pipeline.py (hunk 2 above)
Comment (turian) [reply heareval/tasks/pipeline.py:440]: Fixed

File: heareval/tasks/pipeline.py (hunk 5 above)
Comment (turian) [reply heareval/tasks/pipeline.py:439]: So I can look up in it quickly

**Pull request merged**